spark/src/test/scala/org/apache/spark/sql/delta/SnapshotManagementSuite.scala

```diff
@@ -473,4 +480,84 @@ class SnapshotManagementSuite extends QueryTest with DeltaSQLTestUtils with Shar
       spark.read.format("delta").load(path).collect()
     }
   }
+
+  Seq(true, false).foreach { gapAtTheEnd =>
+    test(s"concurrent backfills are properly reconciled with gapAtTheEnd: $gapAtTheEnd") {
+      val commitStoreName = "awaiting-commit-store"
+      // We defer backfills to manually trigger concurrent backfills for versions [3, 5] during
+      // CommitStore.getCommits but after the LogStore listing.
+      // Depending on the value of gapAtTheEnd, we also write versions [6, 8] to the log.
+      val deferredBackfillMap: mutable.Map[Long, () => Unit] = mutable.Map.empty
+      var finishPendingBackfillsDuringListCommits = false
+
+      case class AwaitingCommitStore() extends InMemoryCommitStore(5) {
+        override def getCommits(
+            logPath: Path,
+            startVersion: Long,
+            endVersion: Option[Long]): GetCommitsResponse = {
+          if (finishPendingBackfillsDuringListCommits) {
+            logInfo(
+              s"Finishing pending backfills concurrently with the list calls " +
+                s"${deferredBackfillMap.keySet}")
+            // Add a 5s delay to ensure file-system listFrom call completes and then backfill
+            // commit [3, 5] to create a perceived gap between file-system deltas and
+            // commit-store commits and test the reconciliation logic works as expected.
+            sleep(5000)
```

Comment:
Could we use semaphores instead of sleep?

Reply:
We could but it would require creating a LogStore wrapper to intercept the List call. Registering the wrapper would need creating a new test suite. Feels like the extra complexity isn't worth it for the test.